{% extends "layout.html" %}

{% block nav_About %} <li class="nav-item active"> <a class="nav-link" href="{{ url_for('about') }}">About<span class="sr-only">(current)</span></a></li> {% endblock %}

{% block content %}
    <br><br>
    <div class="row"><h3>What is aGOtool?</h3>
        <div class="col-md-12">
            We're so glad you asked! Well this is yet another Gene Ontology enrichment tool, BUT with some useful features you'll not find elsewhere. This tool is protein-centric and provides multiple enrichment methods in order to provide biologically meaningful results. We support all UniProt identifiers (SwissProt and TrEMBL) as well as STRING identifiers. You can enrich for <a href="https://www.ebi.ac.uk/QuickGO">GO</a> (molecular function, biological process, cellular component), <a href="https://www.uniprot.org/help/keywords">UniProt keywords</a>, <a href="https://www.genome.jp/kegg/pathway.html">KEGG pathways</a>, <a href="https://www.ncbi.nlm.nih.gov/pubmed">PubMed publications</a>, <a href="https://reactome.org">Reactome</a>, <a href="https://www.wikipathways.org">Wiki Pathways</a>, <a href="https://www.ebi.ac.uk/interpro">Interpro domains</a>, <a href="https://pfam.xfam.org">PFAM domains</a>, <a href="https://www.ebi.ac.uk/ols/ontologies/bto">Brenda Tissues</a>, and <a href="http://disease-ontology.org">Disaeses</a>. We're providing <b>regular monthly updates</b>. Speaking of which we're using textmining results which enables us to provide you with enrichment for Diseases, Tissues, PubMed publications, and Compartments (mapped to and reported as "Gene Ontology Cellular Component TEXTMINING", which is a separate category to "Gene Ontology Cellular Component" stemming from UniProt).
        </div>
    </div>
    <br><br>

    <div class="row"><h3>Support & Questions</h3></div>
    <div class="row"><h5>We welcome your feedback!</h5></div>
    <div class="row">
        <div class="col-md-4">David Lyon</div>
    </div>
    <div class="row">
        <div class="col-md-4"><a href="mailto:david.lyon@uzh.ch" class="link"><span>david.lyon@uzh.ch</span></a></div>
    </div>
    <div class="row">
        <div class="col-md-10"><a href="https://www.imls.uzh.ch/en/research/vonmering.html">Department for Systems Biology / Metagenomics, Institute for Molecular Life Sciences, University of Zurich</a></div>
    </div>
    <br>
{#    <div class="row"><h3>Complaints</h3></div>#}
    <div class="row">
        <div class="col-md-4">Christian von Mering</div>
        <div class="col-md-4">Lars Juhl Jensen</div>
    </div>
    <div class="row">
        <div class="col-md-4"><a href="mailto:mering@imls.uzh.ch" class="link"><span>mering@imls.uzh.ch</span></a></div>
        <div class="col-md-4"><a href="mailto:lars.juhl.jensen@cpr.ku.dk" class="link"><span>lars.juhl.jensen@cpr.ku.dk</span></a></div>
    </div>
    <br><br>

    <div class="row"><h3>Publication</h3></div>
        <div class="row">
            <div class="col-md-12">Please cite the following article when applying this method or using this service:<br>
            <a href="http://www.ncbi.nlm.nih.gov/pubmed/?term=10.1038%2Fnmeth.3621">Avoiding abundance bias in the functional annotation of posttranslationally modified proteins.<br>Nature Methods, 2015<br>doi:10.1038/nmeth.3621</a>
            </div>
        </div>
    <br><br>

    <div class="row"><h3>Source Code</h3></div>
        <div class="row"><div class="col-md-12">The software (published under <a href="http://opensource.org/licenses/BSD-2-Clause"> BSD</a>) is available at <a href="https://github.com/dblyon/agotool.git">https://github.com/dblyon/agotool.git</a>.<br>Feel free to contact <a href="mailto:david.lyon@uzh.ch" class="link"><span>david.lyon@uzh.ch</span></a> for support.</div>
        </div>

    <br><br>


    <div class="row"><h3>Resources</h3></div>
    <div class="row">
        <div class="col-md-12">
            <b>The resources used for this web-service are automatically updated on a monthly basis.</b><br>
            Functional associations for the following categories are retrieved from <a href="https://www.uniprot.org">UniProt</a>: GO terms, UniProt Keywords, KEGG, Interpro, Pfam, and Reactome. Since these associations can simply be counted we apply a Fisher's exact test to find significantly enriched terms.
            <a href="https://diseases.jensenlab.org">Diseases</a>, Brenda <a href="https://tissues.jensenlab.org">Tissues</a>, Gene Ontology Cellular Component (<a href="https://compartments.jensenlab.org">Compartments</a>), and PubMed publications stem from <a href="https://jensenlab.org">Jensenlab</a> textming results. Since these protein to function associations are not binary, but result in a score between 0 and 5, we apply a Kolmogorov Smirnov test.
{#            GO basic and slims subsets are retrieved from <a href="http://purl.obolibrary.org">obolibrary</a>.<br>#}
        </div>
    </div>

<br><br><br><br><br>
{% endblock content %}
